rhp/client.go

```
@@ -0,0 +1,238 @@
+	rev, err := rhp.RPCLatestRevision(ctx, c.client, contractID)
+	if err != nil {
+		return rhp.RPCRenewContractResult{}, fmt.Errorf("failed to fetch latest revision: %w", err)
+	} else if rev.Renewed {
+		return rhp.RPCRenewContractResult{}, fmt.Errorf("contract already renewed")
+	} else if !rev.Revisable {
+		return rhp.RPCRenewContractResult{}, fmt.Errorf("contract not revisable")
+	}
+
+	// NOTE: when renewing a contract we keep the same allowance and collateral.
+	// This has the following advantages:
+	// 1. Contracts drain over time if they contain more funds than needed
+	// 2. Renewals are very "cheap" since no party needs to lock away
+	//    additional funds. Only the fees need to be paid.
+	res, err := rhp.RPCRenewContract(ctx, c.client, c.cm, c.signer, c.cm.TipState(), settings.Prices, rev.Contract, proto.RPCRenewContractParams{
+		ContractID:  contractID,
+		Allowance:   rev.Contract.RenterOutput.Value,
+		Collateral:  rev.Contract.MissedHostValue,
+		ProofHeight: proofHeight,
+	})
+	if err != nil {
+		return rhp.RPCRenewContractResult{}, fmt.Errorf("failed to form contract: %w", err)
+	}
+
+	return res, nil
+}
+
+// ReplenishAccounts replenishes the accounts in the contract to the target value.
+func (c *HostClient) ReplenishAccounts(ctx context.Context, contractID types.FileContractID, accounts []proto.Account, target types.Currency) (res rhp.RPCReplenishAccountsResult, funded int, _ error) {
+	rev, err := c.LatestRevision(ctx, contractID)
+	if err != nil {
+		return rhp.RPCReplenishAccountsResult{}, 0, fmt.Errorf("failed to fetch latest revision: %w", err)
+	} else if rev.Contract.RenterOutput.Value.Cmp(target) < 0 {
+		return rhp.RPCReplenishAccountsResult{}, 0, ErrContractInsufficientFunds
+	}
+
+	// prepare batch
+	batchSize := int(min(rev.Contract.RenterOutput.Value.Div(target).Big().Uint64(), proto.MaxAccountBatchSize))
+	funded = min(batchSize, len(accounts))
+	batch := accounts[:funded]
```

Why not return an error if funded > batch size? This adds an unnecessary extra return value to check.

---

If you look at `Funder.FundAccounts`, my intention was to allow passing in a large number of contracts and accounts to fund. Potentially larger than `proto.MaxAccountBatchSize` since it's a batch size after all. We loop the contracts and pass at most `proto.MaxAccountBatchSize` for every contract. The contract might not have enough funds to fund all accounts, but we do want to drain them as close to zero as possible, so that's why I return the number of funded accounts and limit the batch size in the client to the max accounts that can be funded with that contract. I think this is better than skipping over the contract because we can't fill all accounts. 

Preparing this answer I did have to alter the implementation of `Funder.FundAccounts` though. Refactoring this left an unwanted check that was limiting the number of accounts you can pass to the max batch size.